server.go

```diff
@@ -148,38 +208,69 @@ func (srv *Server) Publish(channels []string, ev Event) {
 	}
 }
 
-// PublishComment publishes a comment to one or more channels.
-func (srv *Server) PublishComment(channels []string, text string) {
+// PublishWithAcknowledgment publishes an event to one or more channels, returning a channel that will receive
+// a value after the event has been processed by the server.
+//
+// This can be used to ensure a well-defined ordering of operations. Since each Server method is handled
+// asynchronously via a separate channel, if you call server.Publish and then immediately call server.Close,
+// there is no guarantee that the server execute the Close operation only after the event has been published.
+// If you instead call PublishWithAcknowledgement, and then read from the returned channel before calling
+// Close, you can be sure that the event was published before the server was closed.
+func (srv *Server) PublishWithAcknowledgment(channels []string, ev Event) <-chan struct{} {
+	ackCh := make(chan struct{}, 1)
 	srv.pub <- &outbound{
 		channels:       channels,
-		eventOrComment: comment{value: text},
+		eventOrComment: ev,
+		ackCh:          ackCh,
 	}
+	return ackCh
 }
 
-func replay(repo Repository, sub *subscription) {
-	for ev := range repo.Replay(sub.channel, sub.lastEventID) {
-		sub.out <- ev
+// PublishComment publishes a comment to one or more channels.
+func (srv *Server) PublishComment(channels []string, text string) {
+	srv.pub <- &outbound{
+		channels:       channels,
+		eventOrComment: comment{value: text},
 	}
 }
 
 func (srv *Server) run() {
+	// All access to the subs and repos maps is done from the same goroutine, so modifications are safe.
 	subs := make(map[string]map[*subscription]struct{})
 	repos := make(map[string]Repository)
+	trySend := func(sub *subscription, ec eventOrComment) {
+		if !sub.send(ec) {
+			sub.close()
+			delete(subs[sub.channel], sub)
```

Comment:
Previously, if we detected that a subscriber had fallen behind and should be removed, we were notifying ourselves of this by pushing to the "unsubscribe" channel (for instance, on line 180 in the old code). But since that channel is only being read in this same loop, and this goroutine owns the `subs` map, there's no reason not to just immediately remove it from the map instead of posting a message to ourselves.